Normative: Add class `static` initialization blocks

rbuckton opened the pull request:
This adds the specification text from https://github.com/tc39/proposal-class-static-block, currently at Stage 3.

jmdyck commented on spec.html:
```diff
@@ -21128,6 +21240,34 @@ <h1>Static Semantics: Early Errors</h1>
+      <emu-grammar>ClassStaticBlockBody: ClassStaticBlockStatementList</emu-grammar>
+      <ul>
+        <li>
+          <p>It is a Syntax Error if the LexicallyDeclaredNames of |ClassStaticBlockStatementList| contains any duplicate entries.</p>
```
```suggestion
          It is a Syntax Error if the LexicallyDeclaredNames of |ClassStaticBlockStatementList| contains any duplicate entries.
```
(and likewise below)

jmdyck commented on spec.html:
```diff
@@ -21410,6 +21558,150 @@ <h1>Static Semantics: ContainsArguments</h1>
+      <emu-alg>
+        1. Return *false*.
+      </emu-alg>
+      <emu-grammar>PropertyName : ComputedPropertyName</emu-grammar>
+      <emu-alg>
+        1. Return the result of ComputedPropertyContainsAwait for |ComputedPropertyName|.
+      </emu-alg>
```
This definition could be dropped, since the chain rule will handle it.

But there needs to be a definition for ComputedPropertyName.

rbuckton replied:
This is a typo. It should be `ContainsAwait for |ComputedPropertyName|`.

jmdyck commented on spec.html:
```diff
@@ -21410,6 +21558,150 @@ <h1>Static Semantics: ContainsArguments</h1>
+      </emu-alg>
+      <emu-grammar>ClassElementList : ClassElementList ClassElement</emu-grammar>
+      <emu-alg>
+        1. Let ComputedPropertyContainsAwait for |ClassElementList| is *true*, return *true*.
```
```suggestion
        1. If ComputedPropertyContainsAwait for |ClassElementList| is *true*, return *true*.
```

jmdyck commented on spec.html:
```diff
@@ -21438,6 +21730,21 @@ <h1>Runtime Semantics: ClassFieldDefinitionEvaluation</h1>
+        1. Let _lex_ be the running execution context's LexicalEnvironment.
+        1. Let _privateScope_ be the running execution context's PrivateEnvironment.
+        1. Let _sourceText_ be the empty sequence of Unicode code points.
+        1. Let _body_ be OrdinaryFunctionCreate(%Function.prototype%, _sourceText_, &laquo; &raquo;, |ClassStaticBlockBody|, ~non-lexical-this~, _lex_, _privateScope_).
```
```suggestion
        1. Let _func_ be OrdinaryFunctionCreate(%Function.prototype%, _sourceText_, &laquo; &raquo;, |ClassStaticBlockBody|, ~non-lexical-this~, _lex_, _privateScope_).
```
(and likewise below)

jmdyck commented on spec.html:
```diff
@@ -21438,6 +21730,21 @@ <h1>Runtime Semantics: ClassFieldDefinitionEvaluation</h1>
+        1. Perform MakeMethod(_body_, _homeObject_).
+        1. Return the ClassStaticBlockDefinition Record { [[Body]]: _body_ }.
+      </emu-alg>
+      <emu-note>The function created for _body_ is never directly accessible to ECMAScript code.</emu-note>
```
```suggestion
      <emu-note>The function _body_ is never directly accessible to ECMAScript code.</emu-note>
```

rbuckton replied:
This text mirrors the text for _ClassFieldDefinitionEvaluation_ just above it.

jmdyck replied:
Ah, yeah, that's wrong too. There isn't a function *created for* `_initializer_`/`_body_`; `_initializer_`/`_body_` *is* the function (that has been created for _Initializer_/_ClassStaticBlockBody_, if you like).

jmdyck commented on spec.html:
```diff
@@ -21057,6 +21160,15 @@ <h2>Syntax</h2>
         PropertyName[?Yield, ?Await]
         PrivateIdentifier
+
+      ClassStaticBlock :
+        `static` `{` ClassStaticBlockBody `}`
+
+      ClassStaticBlockBody :
+        ClassStaticBlockStatementList
+
+      ClassStaticBlockStatementList :
+        StatementList[~Yield, +Await, ~Return]?
```
These need to be referenced from Annex A.

jmdyck reviewed (changes requested):
Also, I believe the following SDO definitions are needed:
- EvaluateBody for ClassStaticBlockBody
- PrivateBoundIdentifiers for ClassStaticBlock
- ComputedPropertyContainsAwait for ClassStaticBlock

rbuckton commented on spec.html:
```diff
@@ -4191,6 +4191,40 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+          </tr>
+          <tr>
+            <td>
+              [[Body]]
```
The parallel I was drawing was to the ClassFieldDefinitionRecord, which has an `[[Initializer]]` slot for the function created to represent the _Initializer_ of a field, as this is the function created to represent the _ClassStaticBlockBody_. `[[Function]]` seems somewhat generic and unclear when you encounter it elsewhere in the specification.

rbuckton commented on spec.html:
```diff
@@ -6489,6 +6540,14 @@ <h1>Static Semantics: LexicallyScopedDeclarations</h1>
       <emu-alg>
         1. Return the TopLevelLexicallyScopedDeclarations of |StatementList|.
       </emu-alg>
+      <emu-grammar>ClassStaticBlockStatementList: [empty]</emu-grammar>
+      <emu-alg>
+        1. Return a new empty List.
+      </emu-alg>
+      <emu-grammar>ClassStaticBlockStatementList: StatementList</emu-grammar>
+      <emu-alg>
+        1. Return the TopLevelLexicallyScopedDeclarations of |StatementList|.
+      </emu-alg>
```
Ah, it looks like I'm missing a definition for `EvaluateBody`, I will add that.

rbuckton commented:
Thanks @jmdyck. I believe I've addressed your feedback, though I chose to use `[[BodyFunction]]` over `[[Function]]` and `_bodyFunction_` over `_func_` to further improve clarity.

jmdyck commented on spec.html:
```diff
@@ -11285,6 +11387,13 @@ <h1>Runtime Semantics: EvaluateBody</h1>
+        <emu-grammar>
+          ClassStaticBlockBody : ClassStaticBlockStatementList
+        </emu-grammar>
+        <emu-grammar>
```
```suggestion
        <emu-alg>
```

jmdyck commented on spec.html:
```diff
@@ -11285,6 +11387,13 @@ <h1>Runtime Semantics: EvaluateBody</h1>
+        <emu-grammar>
+          1. Assert: _argumentsList_ is empty.
+          1. Return ? EvaluateClassStaticBlockBody of |ClassStaticBlockBody| with argument _functionObject_.
+        </emu-grammar>
```
```suggestion
        </emu-alg>
```

jmdyck commented on spec.html:
```diff
@@ -41583,6 +41924,9 @@ <h1>Functions and Classes</h1>
     <emu-prodref name="ClassElementName"></emu-prodref>
+    <emu-prodref name="ClassStaticBlock"></emu-prodref>
+    <emu-prodref name="ClassStaticBlockBody"></emu-prodref>
+    <emu-prodref name="ClassStaticBlockBodyStatementList"></emu-prodref>
```
```suggestion
    <emu-prodref name="ClassStaticBlockStatementList"></emu-prodref>
```

jmdyck commented on spec.html:
```diff
@@ -21438,6 +21743,31 @@ <h1>Runtime Semantics: ClassFieldDefinitionEvaluation</h1>
+      <emu-grammar>ClassStaticBlockBody : ClassStaticBlockStatementList</emu-grammar>
+      <emu-alg>
+        1. Perform ? FunctionDeclarationInstantiation(_functionObject_, &laquo; &raquo;).
+        1. Return the result of evaluating |ClassStaticBlockStatementList|.
```
Evaluation of `ClassStaticBlockStatementList : StatementList` can chain to `StatementList`, but
Evaluation of `ClassStaticBlockStatementList : [empty]` needs to be defined.

rbuckton replied:
I've added this to sec-class-definitions-runtime-semantics-evaluation, thanks.

jmdyck commented on spec.html:
```diff
@@ -5813,6 +5847,15 @@ <h1>InitializeInstanceElements ( _O_, _constructor_ )</h1>
+
+    <emu-clause id="sec-evaluatestaticblock" aoid="EvaluateStaticBlock">
+      <h1>EvaluateStaticBlock ( _receiver_, _blockRecord_ )</h1>
+      <emu-alg>
+        1. Assert: Type(_receiver_) is Object.
+        1. Assert: _blockRecord_ is a ClassStaticBlockDefinition Record.
```
```suggestion
      <p>The abstract operation EvaluateStaticBlock takes arguments _receiver_ (an Object) and _blockRecord_ (a ClassStaticBlockDefinition Record). It performs the following steps when called:</p>
      <emu-alg>
```

jmdyck commented on spec.html:
```diff
@@ -21438,6 +21743,31 @@ <h1>Runtime Semantics: ClassFieldDefinitionEvaluation</h1>
+        1. Let _lex_ be the running execution context's LexicalEnvironment.
+        1. Let _privateScope_ be the running execution context's PrivateEnvironment.
+        1. Let _sourceText_ be the empty sequence of Unicode code points.
+        1. Let _bodyFunction_ be OrdinaryFunctionCreate(%Function.prototype%, _sourceText_, &laquo; &raquo;, |ClassStaticBlockBody|, ~non-lexical-this~, _lex_, _privateScope_).
```
OrdinaryFunctionCreate's 3rd parameter is a Parse Node, not a List.
```suggestion
        1. Let _formalParameters_ be an instance of the production <emu-grammar>FormalParameters : [empty]</emu-grammar>.
        1. Let _bodyFunction_ be OrdinaryFunctionCreate(%Function.prototype%, _sourceText_, _formalParameters_, |ClassStaticBlockBody|, ~non-lexical-this~, _lex_, _privateScope_).
```

rbuckton commented:
@tc39/ecma262-editors would you be able to review this?

jmdyck commented on spec.html:
```diff
@@ -6747,6 +6781,16 @@ <h1>
       </emu-alg>
     </emu-clause>
+
+    <emu-clause id="sec-evaluatestaticblock" aoid="EvaluateStaticBlock">
+      <h1>EvaluateStaticBlock ( _receiver_, _blockRecord_ )</h1>
+      <p>The abstract operation EvaluateStaticBlock takes arguments _receiver_ (an Object) and _blockRecord_ (a ClassStaticBlockDefinition Record). It performs the following steps when called:</p>
```
```suggestion
    <emu-clause id="sec-evaluatestaticblock" type="abstract operation">
      <h1>
        EvaluateStaticBlock (
          _receiver_: an Object,
          _blockRecord_: a ClassStaticBlockDefinition Record,
        )
      </h1>
      <dl class="header">
      </dl>
```

rbuckton replied:
Ah, I didn't realize this was a thing. I'm a bit confused by the `<dl class="header">` tag though. I assume it provides non-printing metadata for the clause? Its a bit confusing to have two metadata formats in the spec HTML:

- `<pre class="metadata">` &mdash; Used at the top for document metadata.
- `<dl class="header">` &mdash; Used for clauses.

Also, It doesn't seem like the documentation at https://tc39.es/ecmarkup/#metadata has been updated to reflect this format.

jmdyck replied:
> I'm a bit confused by the `<dl class="header">` tag though. I assume it provides non-printing metadata for the clause (or somehow attaches extra info to the clause)?

Metadata yes (well, depending on your definition of "metadata"), non-printing no. Currently, I think the content of any `<dd>` that appears in a `dl.header` will also appear in the resulting rendered preamble. But in future, it might also contain non-printing info. 

> Also, It doesn't seem like the documentation at https://tc39.es/ecmarkup/#metadata has been updated to reflect this format.

Yeah, I think the closest it has is [the PR that added support for the new format](https://github.com/tc39/ecmarkup/pull/327).

jmdyck commented on spec.html:
```diff
@@ -6747,6 +6781,16 @@ <h1>
+      <h1>EvaluateStaticBlock ( _receiver_, _blockRecord_ )</h1>
+      <p>The abstract operation EvaluateStaticBlock takes arguments _receiver_ (an Object) and _blockRecord_ (a ClassStaticBlockDefinition Record). It performs the following steps when called:</p>
+      <emu-alg>
+        1. Assert: Type(_receiver_) is Object.
+        1. Assert: _blockRecord_ is a ClassStaticBlockDefinition Record.
```
and delete these.

rbuckton commented:
SpiderMonkey intends to ship this unflagged in 93, and it is already shipping in V8. I'd like to seek Stage 4 at the upcoming meeting, assuming an editor review is complete and approved prior to the agenda submission deadline at the end of the week.

jmdyck commented on spec.html:
```diff
@@ -6757,6 +6791,20 @@ <h1>
+    <emu-clause id="sec-evaluatestaticblock" type="abstract operation">
+      <h1>
+        EvaluateStaticBlock (
+          _receiver_ : an Object,
```
```suggestion
          _receiver_: an Object,
```

syg commented on spec.html:
```diff
@@ -24010,6 +24172,154 @@ <h1>Static Semantics: ContainsArguments</h1>
     </emu-clause>
 
+    <emu-clause id="sec-static-semantics-containsawait" type="sdo" aoid="ContainsAwait">
+      <h1>Static Semantics: ContainsAwait</h1>
```
Aside: it is unfortunate we don't have a predicate-taking version of Contains. If this pattern comes up a few more times maybe we can refactor.

syg replied:
I should say there are quite a few of these ContainsFoo SDOs already, so the "comes up a few more times" threshold might have already been met? And maybe it wouldn't save that much space or readability anyway.

bakkot replied:
In the TLA PR we [decided](https://github.com/tc39/ecma262/pull/2408#discussion_r671918604) we were OK with using ``Contains `await`.`` I still think that's a bit suspicious, but since it's precedented now, I suppose we can use it here instead of the new SDO.

syg replied:
Ah true, I had forgotten that.

rbuckton replied:
As long as this catches `for await` (which I think it does, given the definition of Contains), then I'm happy to remove this.

bakkot commented on spec.html:
```diff
@@ -4585,6 +4585,40 @@ <h1>The ClassFieldDefinition Record Specification Type</h1>
+
+    <emu-clause id="sec-classstaticblockdefinition-record-specification-type">
+      <h1>The ClassStaticBlockDefinition Record Specification Type</h1>
+      <p>A <dfn>ClassStaticBlockDefinition Record</dfn> is a Record value used to encapsulate the executable code for a class static initialization block.</p>
```
```suggestion
      <p>A <dfn variants="ClassStaticBlockDefinition Record">ClassStaticBlockDefinition Record</dfn> is a Record value used to encapsulate the executable code for a class static initialization block.</p>
```

(We [just now](https://github.com/tc39/ecma262/pull/2492) added support for auto-linking plural forms when specified like this.)

bakkot reviewed (approved):
LGTM apart from open comments.
